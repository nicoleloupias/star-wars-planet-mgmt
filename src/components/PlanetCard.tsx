import { Box, Card, Center, Circle, FlexProps, Heading, Text } from "@chakra-ui/react";
import { Link } from "react-router-dom";
import stars from "../assets/stars.png";
import { getRandomPlanetColor } from "../helpers/planet";
import { Planet } from "../hooks/usePlanetsStore";

interface PlanetCardProps extends Omit<FlexProps, "id">, Planet {}

export const PlanetCard = ({ id, name, diameter, climate, terrain, population, url, ...props }: PlanetCardProps) => {
  const planetColor = getRandomPlanetColor(name);

  return (
    <Card
      as={Link}
      to={`/${id}`}
      w={80}
      h={420}
      borderRadius={12}
      overflow="hidden"
      bgColor="transparent"
      _hover={{
        ".planet": {
          width: "400px",
          height: "400px"
        },
        ".cardDetails": {
          backgroundImage: "linear-gradient(180deg, rgba(255,18,73,0.5) 0%, rgba(255,0,0,0)) 70%"
        }
      }}
      borderWidth="1px"
      borderColor={"rgba(255, 18, 73, 0.2)"}
      {...props}
    >
      <Center overflow="hidden" h="224px" bgImage={`url(${stars})`} bgSize="200px">
        <Circle
          className="planet"
          mx="auto"
          size={40}
          bgColor={planetColor}
          boxShadow={`0px 0px 30px 0px ${planetColor}`}
          transition="0.5s ease-out all"
        />
      </Center>

      <Box
        px={4}
        py={4}
        backgroundImage="linear-gradient(180deg, rgba(255,18,73,0.3) 0%, rgba(255,0,0,0)) 70%"
        minH={196}
        color="gray.200"
        position="absolute"
        bottom={0}
        w="full"
        className="cardDetails"
      >
        <Heading color="gray.50">{name}</Heading>
        <Text>Diameter: {diameter}</Text>
        <Text>Climate: {climate}</Text>
        <Text>Terrain: {terrain}</Text>
        <Text>Population: {population}</Text>
      </Box>
    </Card>
  );
};
