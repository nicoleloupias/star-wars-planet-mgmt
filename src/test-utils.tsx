import { ChakraProvider } from "@chakra-ui/react";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import React from "react";
import { Router, RouterProvider } from "react-router-dom";
import { theme } from "./theme";
const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <ChakraProvider theme={theme}>
      <Router location={history.location} navigator={history}>
        {children}
      </Router>
    </ChakraProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {}
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => <AllProviders history={history}>{children}</AllProviders>,
    ...options
  });

export { customRender as render };
