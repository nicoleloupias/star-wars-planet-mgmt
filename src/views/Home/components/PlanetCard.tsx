import type { FlexProps } from "@chakra-ui/react";
import { Box, Card, Center, Circle, Heading, Text } from "@chakra-ui/react";
import { Link } from "react-router-dom";
import stars from "../../../assets/stars.png";
import { getRandomPlanetColor } from "../../../utils/planet";
import type { Planet } from "../../../hooks/usePlanetsStore";

export interface PlanetCardProps
  extends Omit<FlexProps, "id">,
    Pick<Planet, "name" | "climate" | "diameter" | "terrain" | "population" | "id"> {}

export const PlanetCard = ({ id, name, diameter, climate, terrain, population, ...props }: PlanetCardProps) => {
  const planetColor = getRandomPlanetColor(name);

  const getCircleSize = () => {
    if (diameter === "unknown" || !parseInt(diameter)) return "160px";
    const maxDiameter = 200;
    const minDiameter = 40;
    const scaledDiameter = (parseInt(diameter) / 10000) * maxDiameter;
    const adjustedDiameter = Math.max(Math.min(scaledDiameter, maxDiameter), minDiameter);

    return `${adjustedDiameter}px`;
  };

  return (
    <Card
      as={Link}
      to={`/${id}`}
      w={80}
      h={420}
      borderRadius={30}
      overflow="hidden"
      bgColor="transparent"
      _hover={{
        ".planet": {
          width: "400px",
          height: "400px"
        },
        ".cardDetails": {
          backgroundImage: "linear-gradient(180deg, rgba(255,18,73,0.5) 0%, rgba(255,0,0,0)) 70%"
        }
      }}
      borderWidth="1px"
      borderColor={"rgba(255, 18, 73, 0.2)"}
      data-testid="planet-card"
      {...props}
    >
      <Center overflow="hidden" h="224px" bgImage={`url(${stars})`} bgSize="200px">
        <Circle
          className="planet"
          mx="auto"
          size={getCircleSize()}
          bgColor={planetColor}
          boxShadow={`0px 0px 30px 0px ${planetColor}`}
          transition="0.5s ease-out all"
        />
      </Center>

      <Box
        px={4}
        py={4}
        backgroundImage="linear-gradient(180deg, rgba(255,18,73,0.3) 0%, rgba(255,0,0,0)) 70%"
        minH={196}
        color="gray.200"
        position="absolute"
        bottom={0}
        w="full"
        className="cardDetails"
        textTransform="capitalize"
      >
        <Heading color="gray.50" as="h3" fontWeight={"800"}>
          {name}
        </Heading>
        <Text>
          Diameter: {diameter}{" "}
          {diameter !== "unknown" && (
            <Box as="span" textTransform="lowercase">
              km
            </Box>
          )}
        </Text>
        <Text>Climate: {climate}</Text>
        <Text>Terrain: {terrain}</Text>
        <Text>Habitants: {population}</Text>
      </Box>
    </Card>
  );
};
