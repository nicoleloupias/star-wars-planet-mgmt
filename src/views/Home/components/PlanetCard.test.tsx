import { describe, expect, test } from "vitest";
import { render } from "../../../test-utils";
import { PlanetCard, PlanetCardProps } from "./PlanetCard";

describe("<PlanetCard/>", () => {
  const defaultProps = {
    id: "12",
    name: "Utapau",
    rotation_period: "27",
    orbital_period: "351",
    diameter: "12900",
    climate: "temperate, arid, windy",
    gravity: "1 standard",
    terrain: "scrublands, savanna, canyons, sinkholes",
    surface_water: "0.9",
    population: "95000000",
    residents: ["https://swapi.dev/api/people/83/"],
    films: ["https://swapi.dev/api/films/6/"],
    created: "2014-12-10T12:49:01.491000Z",
    edited: "2014-12-20T20:58:18.439000Z",
    url: "https://swapi.dev/api/planets/12/"
  };
  const factoryComponent = (props: PlanetCardProps = defaultProps) => render(<PlanetCard {...props} />);

  test("should render succesfully", () => {
    const component = factoryComponent();

    expect(component).toMatchSnapshot();
    expect(component.getByTestId("planet-card")).toBeInTheDocument();
  });

  test("should render correctly with default props", () => {
    const component = factoryComponent();

    expect(component.getByText(defaultProps.name)).toBeInTheDocument();
  });

  test("should render correctly with given props", () => {
    const component = factoryComponent({ ...defaultProps, name: "Example" });

    expect(component.getByText("Example")).toBeInTheDocument();
  });

  test("card should be rendered as link", () => {
    const component = factoryComponent({ ...defaultProps, name: "Example" });

    expect(component.getByRole("link")).toBeInTheDocument();
    expect(component.getByRole("link").getAttribute("href")).toBe(`/${defaultProps.id}`);
  });
});
