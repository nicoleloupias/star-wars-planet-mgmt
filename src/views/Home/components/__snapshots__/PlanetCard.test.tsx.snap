// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PlanetCard/> > should render succesfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div>
      <a
        class="chakra-card css-1olgpje"
        created="2014-12-10T12:49:01.491000Z"
        data-testid="planet-card"
        edited="2014-12-20T20:58:18.439000Z"
        films="https://swapi.dev/api/films/6/"
        gravity="1 standard"
        href="/12"
        orbital_period="351"
        residents="https://swapi.dev/api/people/83/"
        rotation_period="27"
        surface_water="0.9"
        url="https://swapi.dev/api/planets/12/"
      >
        <div
          class="css-upfzb1"
        >
          <div
            class="planet css-1mm42zb"
          />
        </div>
        <div
          class="cardDetails css-vk61ae"
        >
          <h2
            class="chakra-heading css-q8veze"
          >
            Utapau
          </h2>
          <p
            class="chakra-text css-0"
          >
            Diameter: 
            12900
            km
          </p>
          <p
            class="chakra-text css-0"
          >
            Climate: 
            temperate, arid, windy
          </p>
          <p
            class="chakra-text css-0"
          >
            Terrain: 
            scrublands, savanna, canyons, sinkholes
          </p>
          <p
            class="chakra-text css-0"
          >
            Population: 
            95000000
          </p>
        </div>
      </a>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <a
      class="chakra-card css-1olgpje"
      created="2014-12-10T12:49:01.491000Z"
      data-testid="planet-card"
      edited="2014-12-20T20:58:18.439000Z"
      films="https://swapi.dev/api/films/6/"
      gravity="1 standard"
      href="/12"
      orbital_period="351"
      residents="https://swapi.dev/api/people/83/"
      rotation_period="27"
      surface_water="0.9"
      url="https://swapi.dev/api/planets/12/"
    >
      <div
        class="css-upfzb1"
      >
        <div
          class="planet css-1mm42zb"
        />
      </div>
      <div
        class="cardDetails css-vk61ae"
      >
        <h2
          class="chakra-heading css-q8veze"
        >
          Utapau
        </h2>
        <p
          class="chakra-text css-0"
        >
          Diameter: 
          12900
          km
        </p>
        <p
          class="chakra-text css-0"
        >
          Climate: 
          temperate, arid, windy
        </p>
        <p
          class="chakra-text css-0"
        >
          Terrain: 
          scrublands, savanna, canyons, sinkholes
        </p>
        <p
          class="chakra-text css-0"
        >
          Population: 
          95000000
        </p>
      </div>
    </a>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
